solutions/usaco-1040.mdx

```diff
@@ -0,0 +1,88 @@
+---
+id: usaco-1040
+source: USACO Silver 2021 Open
+title: The Moo Particle
+author: Mythreya Dharani
+---
+## Explanation
+Let's assume we have a bidirectional graph where any two particles that can interact are connected. 
+We can make the claim that the answer is the number of connected components $c$ in our graph. 
+This is true because we can reduce every connected component to one node in the graph by removing all nodes, moving inwards until we reach a single chosen node to keep.
+
+However, there are at most $10^5$ particles, and therefore, we need a way to calculate $c$ in less than $\mathcal{O}(N^2)$. 
+We can make a $\mathcal{O}(N \log N)$ algorithm by sorting every particle in ascending order, first by its $x$ spin, and then by its $y$ spin, resulting in a list $p_1,  p_2.. p_{N-1},  p_{N}$.
+Each element is intially its own connected component. From there, for each adjacent connected component, as long as the minimum y-coordinate in the left component is at most the maximum y-coordinate of the right component, we combine them together.
+
+For some intution on this algorithm's correctness, since the list is already sorted by x-coordinates, by comparing the minimum of the component with smaller y-coordinates and the maximum of the component with greater y-coordinates, we can uniquely determine whether or not they belong to the same connected component.
```

Comment:
```suggestion
For some intution on this algorithm's correctness, since the list is already sorted by x-coordinates, by comparing the minimum of the component with smaller y-coordinates and the maximum of the component with greater y-coordinates, we can determine whether or not they belong to the same connected component.
```